### PR TITLE
Change BulkCheckPermission response to match order of the request

### LINKED ISSUE #1563: `BulkCheckPermission` does not return results in the same order as requests
### What platforms are affected?

linux, macos, windows

### What architectures are affected?

amd64, arm64

### Steps to Reproduce

I have noticed that in some cases, `BulkCheckPermission` returns the results (permissions) in a different order than requests.

### Expected Result

I think the desired behavior would be to return results in the same order as requests

### Actual Result

-

### PR DESCRIPTION
**Pull request by josephschorr**
Fixes #1563

### CODE REVIEW

File: internal/services/v1/experimental.go [hunk 1]
```diff
@@ -520,11 +533,30 @@ func (es *experimentalServer) BulkCheckPermission(ctx context.Context, req *v1.B
...
+
+		pairHashes[pair] = pairItemHash
+	}
+
+	// Resort the results based on the incoming request order to make it easier to consume. This is not strictly
+	// necessary as each result contains the request that made it, but for the non-streaming version
+	// of this API, it makes sense.
+	slices.SortStableFunc(pairs, func(first *v1.BulkCheckPermissionPair, second *v1.BulkCheckPermissionPair) int {
+		return itemIndexByHash[pairHashes[first]] - itemIndexByHash[pairHashes[second]]
+	})
```
Comment (bradengroom) [inline internal/services/v1/experimental.go:557]: I assume this is going to be an O(n log n) sort, right? You could do the sorting in linear time by re-using your map from above.

```
sortedPairs := make([]*v1.BulkCheckPermissionPair, 0, len(req.Items))
for _, pair := range pairs {
    currHash := computeBulkCheckPermissionItemHash(pair.Request)
    pairIdx := itemIndexByHash[currHash]
    sortedPairs[pairIdx] = pair
}
```

Comment (bradengroom) [reply internal/services/v1/experimental.go:557]: Or if you don't want the added memory of another `sortedPairs` list, you could compute the hash and insert into `pairs` at the proper index in the first loop.

**Review by bradengroom — CHANGES_REQUESTED** [review]
Thanks for doing this. I think you can actually support it without changing the runtime complexity of the endpoint 🙂

File: internal/services/v1/experimental.go (hunk 1 above)
Comment (josephschorr) [reply internal/services/v1/experimental.go]: Updated. Thanks for the suggestion

**Review by bradengroom — APPROVED** [review]
I can't speak to the failing tests, but I'm approving because my feedback was addressed. Thanks!

File: internal/services/v1/experimental.go [hunk 2]
```diff
@@ -520,11 +535,32 @@ func (es *experimentalServer) BulkCheckPermission(ctx context.Context, req *v1.B
...
+	// Run the checks in parallel.
 	if err := tr.StartAndWait(); err != nil {
 		return nil, es.rewriteError(ctx, err)
 	}
 
-	return resp, nil
+	// Resort the results based on the incoming request order to make it easier to consume. This is not strictly
+	// necessary as each result contains the request that made it, but for the non-streaming version
+	// of this API, it makes sense.
+	sortedPairs := make([]*v1.BulkCheckPermissionPair, len(req.Items))
```
Comment (vroldanbet) [inline internal/services/v1/experimental.go:546]: super nit but perhaps we could order the `pairs` slice in-place so we don't allocate another one?

**Review by vroldanbet — APPROVED** [review]

File: internal/services/v1/experimental.go (hunk 2 above)
Comment (josephschorr) [reply internal/services/v1/experimental.go]: Done

**Review by vroldanbet — APPROVED** [review]

**Pull request merged**